### PR TITLE
UI masks per mesh instance

### PR DESCRIPTION
**Pull request by daredevildave**
Add "mask" property to image elements which disables visible rendering of the image element and instead masks child elements by the rect (if no texture) or alpha channel (if textured). Masked elements are only visible inside the masked area.

### CODE REVIEW

File: src/framework/components/element/system.js [hunk 1]
```diff
@@ -45,6 +48,28 @@ pc.extend(pc, function () {
         this.defaultImageMaterial.depthWrite = false;
         this.defaultImageMaterial.update();
 
+        this.defaultImageMaskMaterial = this.defaultImageMaterial.clone();
+        this.defaultImageMaskMaterial.diffuse = new pc.Color(0,0,0,1); // black diffuse color to prevent ambient light being included
+        this.defaultImageMaskMaterial.emissive = new pc.Color(0.5,0.5,0.5,1); // use non-white to compile shader correctly
```
Comment (willeastcott) [inline src/framework/components/element/system.js]: You can just use .set to set this properties. Otherwise you're just 'leaking' a pc.Color object. Plus you only need to set 3 values. Diffuse and emissive properties are RGB, not RGBA.

File: src/framework/components/element/system.js [hunk 2]
```diff
@@ -63,6 +88,29 @@ pc.extend(pc, function () {
         this.defaultScreenSpaceImageMaterial.depthWrite = false;
         this.defaultScreenSpaceImageMaterial.update();
 
+        this.defaultScreenSpaceImageMaskMaterial = this.defaultScreenSpaceImageMaterial.clone();
+        this.defaultScreenSpaceImageMaskMaterial.diffuse = new pc.Color(0,0,0,1); // black diffuse color to prevent ambient light being included
+        this.defaultScreenSpaceImageMaskMaterial.emissive = new pc.Color(0.5,0.5,0.5,1); // use non-white to compile shader correctly
```
Comment (willeastcott) [inline src/framework/components/element/system.js]: Use .set(r, g, b)

File: src/framework/components/element/system.js [hunk 3]
```diff
@@ -225,12 +288,28 @@ pc.extend(pc, function () {
...
+            // var parent = component.entity.getParent();
+            // if (parent && parent.element) {
+            //     if (parent.element.mask) {
+            //         component._updateMask(parent);
+            //     } else {
+            //         if (parent.element._maskEntity) {
+            //             component._updateMask(parent.element._maskEntity);
+            //         }
+            //     }
+            // }
```
Comment (willeastcott) [inline src/framework/components/element/system.js]: If you don't need this code, delete it.

File: src/framework/components/element/text-element.js [hunk 4]
```diff
@@ -225,6 +232,47 @@ pc.extend(pc, function () {
...
+
+                for (var i = 0, len = this._model.meshInstances.length; i<len; i++) {
+                    var mi = this._model.meshInstances[i];
+                    mi.stencilFront = mi.stencilBack = this._maskParams;
+                }
+                this._maskedBy = mask;
+            } else {
+                // remove mask
+                if (this._maskedBy) {
+                    for (var i = 0, len = this._model.meshInstances.length; i<len; i++) {
```
Comment (willeastcott) [inline src/framework/components/element/text-element.js]: Redeclaration of i. Put at start of function.

File: src/framework/components/element/text-element.js [hunk 5]
```diff
@@ -225,6 +232,47 @@ pc.extend(pc, function () {
...
+                        var mi = this._model.meshInstances[i];
```
Comment (willeastcott) [inline src/framework/components/element/text-element.js]: Redeclaration of mi. Put at start of function.

File: src/framework/components/element/image-element.js [hunk 6]
```diff
@@ -231,6 +261,110 @@ pc.extend(pc, function () {
...
+                    var mi = this._model.meshInstances[i];
+                    mi.stencilFront = mi.stencilBack = sp;
+                }
+
+                this._maskedBy = mask;
+            } else {
+                // remove mask
+                // restore default material
+                for (var i = 0, len = this._model.meshInstances.length; i<len; i++) {
+                    var mi = this._model.meshInstances[i];
```
Comment (willeastcott) [inline src/framework/components/element/image-element.js]: Redeclaration of i and mi. Put at start of function.

File: src/framework/components/element/image-element.js [hunk 7]
```diff
@@ -94,18 +107,35 @@ pc.extend(pc, function () {
...
+                   this._material !== this._system.defaultImageMaterial &&
+                   this._material !== this._system.defaultImageMaskMaterial &&
+                   this._material !== this._system.defaultScreenSpaceImageMaskMaterial);
         },
 
+        // assign a material internally without updating everything
+        // _setMaterial: function (material) {
+        //     this._material = material;
+        //     this._meshInstance.material = material;
+        // },
```
Comment (willeastcott) [inline src/framework/components/element/image-element.js]: Delete commented code.

File: src/framework/components/element/image-element.js [hunk 8]
```diff
@@ -1,4 +1,6 @@
 pc.extend(pc, function () {
+    var maskCounter = 1;
```
Comment (willeastcott) [inline src/framework/components/element/image-element.js]: Will this break down with multiple engine instances?

File: src/framework/components/element/component.js [hunk 9]
```diff
@@ -365,12 +372,113 @@ pc.extend(pc, function () {
...
+                    this._image._meshInstance.stencilFront = sp;
+                    this._image._meshInstance.stencilBack = sp;
+                    mask = this.entity;
+                }
+
+                this._maskEntity = null;
+
+                // recurse through all children
+                var children = this.entity.getChildren();
+                for (var i = 0, l = children.length; i < l; i++) {
```
Comment (willeastcott) [inline src/framework/components/element/component.js]: Redeclaration of i. Put at start of function.

File: src/framework/components/element/component.js [hunk 10]
```diff
@@ -365,12 +372,113 @@ pc.extend(pc, function () {
...
+                    this._text._setMaskedBy(null);
+                }
+                if (this._image) {
+                    this._image._setMaskedBy(null);
+                }
+
+                // if this is mask we still need to mask children
+                if (this.mask) {
+                    this._image._maskRef = this._getMaskDepth();
+                    var sp = new pc.StencilParameters({
```
Comment (willeastcott) [inline src/framework/components/element/component.js]: Redeclaration of sp. Put at start of function.

File: src/framework/components/element/component.js [hunk 11]
```diff
@@ -365,12 +372,113 @@ pc.extend(pc, function () {
...
+
+                    this._image._meshInstance.stencilFront = sp;
+                    this._image._meshInstance.stencilBack = sp;
+                    mask = this.entity;
+                }
+
+                this._maskEntity = null;
+
+                // recurse through all children
+                var children = this.entity.getChildren();
```
Comment (willeastcott) [inline src/framework/components/element/component.js]: Redeclaration of children. Put at start of function.

File: examples/masking/index.html [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+                showMenu();
+            }
+
+            var scrolls = [50,50];
+            var a = 0;
+            var SCROLL_AMOUNT = 0.5*((16*64)-300)+10;
+            // update every frame
+            app.on("update", function (dt) {
+
+            });
```
Comment (willeastcott) [inline examples/masking/index.html]: Why is this here? It's empty.

File: src/framework/components/element/component.js [hunk 13]
```diff
@@ -365,12 +372,113 @@ pc.extend(pc, function () {
...
+
+        // set the mask ancestor on this entity
+        _updateMask: function (mask) {
+            if (mask) {
+                var material;
+
+                this._maskEntity = mask;
+
+                if (this._text) {
+                    this._text._setMaskedBy(mask);
```
Comment (guycalledfrank) [inline src/framework/components/element/component.js]: Two different _setMaskedBy should be probably a single function
